### ISSUE TITLE
[Bug]: gpu_memory_utilization affects generation quality

### ISSUE THREAD

**Issue by 20171130** [vllm-project/vllm#15763]
### Your current environment

vllm 0.7.3
python 3.11.5
torch 2.5.1
transformers 4.50.1
cuda/12.6

Using an Nvidia A100 with 40GB memory.

### 🐛 Describe the bug

The parameter "gpu_memory_utilization" significantly affects the quality of generated response.  

mem_limit=0.2, pass@ 1 0.05 2 0.09 4 0.14 8 0.22
mem_limit=0.95, pass@ 1 0.11 2 0.20 4 0.33 8 0.52 

```python
import os
import argparse
from tqdm import tqdm
import torch
import re
from vllm import LLM, SamplingParams
from datasets import load_dataset, Dataset
from peft import PeftModel
import datasets
from transformers import BitsAndBytesConfig, AutoModelForCausalLM, AutoTokenizer, GenerationConfig
import math
from vllm.lora.request import LoRARequest

parser = argparse.ArgumentParser()
parser.add_argument('--precision', type=int, default=16)
parser.add_argument('--greedy', type=bool, default=False, action=argparse.BooleanOptionalAction)
parser.add_argument('--temperature', type=float, default=0.7)
parser.add_argument('--model', type=str, default="Qwen/Qwen2.5-3B-Instruct")
parser.add_argument('--max_len', type=int, default=2048)
parser.add_argument('--lora', type=str, default=None)
parser.add_argument('--mem_limit', type=float)


args = parser.parse_args()
print(args)

def get_knk_dataset(split="train", ppl=[]):
    def extract_answer(x):
        pattern = "(\w+) is a (\w+)"
        pattern = re.compile(pattern)
        matches = pattern.finditer(x)
        result = ''
        for match in matches:
            result += f'{match.groups()[0]}: {match.groups()[1]}\n'
        return result
    prompt = "Think step by step by considering whether each person is lying and if that leads to contradiction. Enclose all reasoning steps within a pair of <reason> tags, then format your answer as '<name>: knight|knave\n', maintaining the order of names as they appear in the question."
    dataset = load_dataset("K-and-K/knights-and-knaves", split)
    if len(ppl) == 0:
        lst = [dataset[item] for item in dataset]
    else:
        lst = [dataset[f'{n}ppl'] for n in ppl]
    dataset = datasets.concatenate_datasets(lst)
    def process(x):
        result = {"prompt": [{"role": "user", "content": x["quiz"]+prompt}]}
        result["completion"] = extract_answer(x["solution_text_format"])
        cot = "<reason>\n" + x["cot_head"] + "\n".join(x["cot_repeat_steps"]) + x["cot_foot"] + "\n</reason>"
        result['cot'] = cot
        return result
    dataset = dataset.map(process, remove_columns=["quiz", "knight_knave", "solution", "solution_text", "solution_text_format", "statements", "cot_repeat_steps", "cot_head", "cot_foot"])
    return dataset

eval_dataset = get_knk_dataset("test", ppl=[]).shuffle(seed=42).select(list(range(100)))
eval_batch_size = 128
eval_shots = 8
max_completion_length = args.max_len
eval_dataset = eval_dataset.batch(batch_size=eval_batch_size//eval_shots)

torch_dtype = torch.bfloat16
quant_storage_dtype = torch.bfloat16

if args.precision==4:
    quantization_config = BitsAndBytesConfig(
            load_in_4bit=True,
            bnb_4bit_use_double_quant=True,
            bnb_4bit_quant_type="nf4",
            bnb_4bit_compute_dtype=torch.bfloat16
        )
elif args.precision == 8:
    quantization_config = BitsAndBytesConfig(
            load_in_8bit=True
        )
else:
    quantization_config = None

model_name = args.model
tokenizer_name = model_name
tokenizer = AutoTokenizer.from_pretrained(tokenizer_name)
tokenizer.pad_token = tokenizer.eos_token
    
if args.precision == 4:
    llm = LLM(
        model=model_name,
        tokenizer=tokenizer_name,
        dtype=torch_dtype,
        enable_prefix_caching=True,
        gpu_memory_utilization=args.mem_limit,
        enable_lora = not(args.lora is None),
        quantization="bitsandbytes", load_format="bitsandbytes"
        )
else:
    llm = LLM(
        model=model_name,
        tokenizer=tokenizer_name,
        dtype=torch_dtype,
        gpu_memory_utilization=args.mem_limit,
        max_model_len=2500,
        enable_prefix_caching=True,
        enable_lora = not(args.lora is None)
        )

if args.greedy:
    args.temperature = 0
lora_request = None
if args.lora is not None:
    lora_request = LoRARequest("lora", 1, args.lora)
sampling_params = SamplingParams(
    temperature=args.temperature,
    max_tokens=max_completion_length,)

def evaluate():
    def extract_answer(x):
        pattern = "(\w+): (\w+)"
        pattern = re.compile(pattern)
        matches = pattern.finditer(x)
        result = {}
        for match in matches:
            result[match.groups()[0]] = match.groups()[1]
        return result

    total = 0
    too_long = 0
    pass_prob = {1:0, 2:0, 4:0, 8:0}
    
    pbar = tqdm(eval_dataset)
    for item in pbar:
        prompts = item["prompt"]
        prompts = tokenizer.apply_chat_template(prompts, tokenize = False, add_generation_prompt = True)
        inputs = []
        for prompt in prompts:
            inputs += [prompt + "<reason>"] * eval_shots
        outputs = llm.generate(inputs, sampling_params=sampling_params, use_tqdm=False, lora_request=lora_request)
        finish_reasons = [output.outputs[0].finish_reason for output in outputs]
        outputs = [output.outputs[0].text for output in outputs]
        too_long += sum([item=="length" for item in finish_reasons])
        for i in range(len(outputs)//eval_shots):
            correct = 0
            for j in range(eval_shots):
                if extract_answer(item["completion"][i]) == extract_answer(outputs[i*eval_shots+j]):
                    correct += 1
            for key in pass_prob:
                if key > eval_shots:
                    continue
                pass_prob[key] += (1-math.comb(eval_shots-correct, key)/math.comb(eval_shots, key))
            total += 1
        pbar.set_description(f"pass@ 1 {pass_prob[1]/total:.2f} 2 {pass_prob[2]/total:.2f} 4 {pass_prob[4]/total:.2f} 8 {pass_prob[8]/total:.2f} too long {too_long/total/eval_shots:.2f}")   

evaluate()
```

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

**Comment by 20171130** [comment]
The accuracy drop is correleted with "WARNING 04-03 22:02:12 scheduler.py:1754] Sequence group 127 is preempted by PreemptionMode.RECOMPUTE mode because there is not enough KV cache space. This can affect the end-to-end performance. Increase gpu_memory_utilization or tensor_parallel_size to provide more KV cache memory." 
It is very likely that the preemption implementation is causing the problem.

**Comment by qinchuanhui** [comment]
I have the same issue in v0.8.0, but i didn't see this warning message.
